packages/notifier/src/notifier.js

```diff
@@ -91,82 +94,82 @@ export const makeNotifierKit = (...args) => {
 
   const final = () => currentUpdateCount === undefined;
 
-  if (args.length >= 1) {
-    // start as hasState() && !final()
-    currentResponse = harden({
-      value: args[0],
-      updateCount: currentUpdateCount,
-    });
-  }
-  // else start as !hasState() && !final()
-
-  // NaN matches nothing
-  function getUpdateSince(updateCount = NaN) {
-    if (
-      hasState() &&
-      (final() ||
-        (currentResponse && currentResponse.updateCount !== updateCount))
-    ) {
-      // If hasState() and either it is final() or it is
-      // not the state of updateCount, return the current state.
-      return Promise.resolve(currentResponse);
-    }
-    // otherwise return a promise for the next state.
-    assert(nextPromiseKit);
-    return nextPromiseKit.promise;
-  }
+  const baseNotifier = harden({
+    // NaN matches nothing
+    getUpdateSince(updateCount = NaN) {
+      if (
+        hasState() &&
+        (final() ||
+          (currentResponse && currentResponse.updateCount !== updateCount))
+      ) {
+        // If hasState() and either it is final() or it is
+        // not the state of updateCount, return the current state.
+        return Promise.resolve(currentResponse);
+      }
+      // otherwise return a promise for the next state.
+      assert(nextPromiseKit);
+      return nextPromiseKit.promise;
+    },
+  });
 
-  function updateState(state) {
-    if (final()) {
-      throw new Error('Cannot update state after termination.');
-    }
-
-    // become hasState() && !final()
-    assert(nextPromiseKit && currentUpdateCount);
-    currentUpdateCount += 1;
-    currentResponse = harden({
-      value: state,
-      updateCount: currentUpdateCount,
-    });
-    nextPromiseKit.resolve(currentResponse);
-    nextPromiseKit = producePromise();
-  }
+  const asyncIterable = makeAsyncIterableFromNotifier(baseNotifier);
 
-  function finish(finalState) {
-    if (final()) {
-      throw new Error('Cannot finish after termination.');
-    }
-
-    // become hasState() && final()
-    assert(nextPromiseKit);
-    currentUpdateCount = undefined;
-    currentResponse = harden({
-      value: finalState,
-      updateCount: currentUpdateCount,
-    });
-    nextPromiseKit.resolve(currentResponse);
-    nextPromiseKit = undefined;
-  }
+  const notifier = harden({
+    ...baseNotifier,
+    ...asyncIterable,
+  });
 
-  function reject(reason) {
-    if (final()) {
-      throw new Error('Cannot reject after termination.');
-    }
+  const updater = harden({
+    updateState(state) {
+      if (final()) {
+        throw new Error('Cannot update state after termination.');
+      }
+
+      // become hasState() && !final()
+      assert(nextPromiseKit && currentUpdateCount);
```

Comment:
It would be better to use `assert(hasState && !final())`. Why leave it this way now?

Reply:
The assertions are to guide Typescript's inference.  It is not smart enough to inline the effects from `hasState() && !final()` to infer that `nextPromiseKit` and `currentUpdateCount` are not undefined.

Reply:
sigh. I wish we didn't need to write worse code in order to satisfy the type checker. But such is life with static checkers.